Sprint 19

Lisiykod opened the pull request:
Привет!

Lisiykod committed:
сверстана карточка для выбора направления + добавлены новые файлы + цвета в ассетах

Lisiykod committed:
сверстан экран настроек

Lisiykod committed:
реализована начальная верстка для экранов со списком городов, станций и списка перевозчиков.
добавлены методы получения городов и станций во вью модель

Lisiykod committed:
редактирование верстки и инициализации готовых вью, сверстан экран с ошибками и карточкой перевозчика

Lisiykod committed:
навигация перенесена в сервис, скорректирован метод поиска и ямл файл (не все данные правильно парсились).
скорректировано отображение карточки перевозчика, списка перевозчиков, серч бара, поиска по городам и станциям, и карточки с выбором направления

Lisiykod committed:
добавлен экран фильтров, новые методы во вью модель + небольшая коррекция верстки

Lisiykod committed:
реализована темная тема по переключению кнопки "темная тема" в настройках,
реализован экран с пользовательским соглашением,
устранены баги отображения пустого результата поиска +
небольшие улучшения кода.

Lisiykod committed:
скорректировано отображение шеврона

NaR0RoW commented on TravelSchedule/Services/NavigationService.swift:
```diff
@@ -0,0 +1,29 @@
+
+import Foundation
+
+class NavigationService: ObservableObject {
```
Стоит пометить как `final`. Если класс не наследуется, то лучше пометить его как `final` для ускорения работы компилятора

NaR0RoW replied:
Обычно такие классы называют [Router](https://www.kodeco.com/books/design-patterns-by-tutorials/v3.0/chapters/23-coordinator-pattern)

NaR0RoW commented on TravelSchedule/Services/NavigationService.swift:
```diff
@@ -0,0 +1,29 @@
+
+class NavigationService: ObservableObject {
+    
+    static let shared = NavigationService()
```
Если данные влияют на UI и должны обновляться во всем приложении, то стоит использовать [EnvironmentObject](https://developer.apple.com/documentation/swiftui/environmentobject)

NaR0RoW commented on TravelSchedule/TravelScheduleApp.swift:
```diff
@@ -9,9 +9,15 @@ import SwiftUI
 struct TravelScheduleApp: App {
+    
+    @State private var viewModel = ScheduleViewModel()
+    @AppStorage("appTheme") private var isDarkModeOn = false
```
Строковое представление можно в отдельную сущность вынести

NaR0RoW commented on TravelSchedule/Services/NavigationService.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    @Published var path: [Route] = []
+    
+    private init() {}
```
В `SwiftUI` лучше не использовать синглтоны при работе с UI

NaR0RoW commented on TravelSchedule/TravelScheduleApp.swift:
```diff
@@ -9,9 +9,15 @@ import SwiftUI
 @main
 struct TravelScheduleApp: App {
+    
+    @State private var viewModel = ScheduleViewModel()
```
`@State` предназначен для управления локальным состоянием View, а не для хранения `ViewModel`. Тут стоит использовать `@StateObject` или `@EnvironmentObject`

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+    
+    @Published var allSettlements: [Settlements] = []
+    @Published var stations: [Stations] = []
+    @Published var fromSettlemet: Settlements?
```
Ошибка в правописании:
`fromSettlemet` -> `fromSettlement`

Лучше включить [настройку](https://sarunw.com/posts/spell-checking-in-xcode/), чтобы `Xcode` подсвечивал ошибки в правописании

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+    func setSettlementsStations(on settlement: Settlements, direction: Direction) {
+        let allStations = settlement.stations ?? []
+        stations = allStations.filter { $0.station_type == "train_station" || $0.transport_type == "train" }
+        stations.sort {$0.title! < $1.title! }
```
Не стоит использовать `force` операции

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+    @Published var isFilter: Bool = false
+    @Published var filteredCarriersList: [Segments] = []
+    @Published var departureTimeIntervals: [DepartureTimeInterval] = []
+    @Published var hasTransfers: Bool = true
```
Лучше не делать View Model с таким большим количеством `@Published` полей, а разбивать модель на подмодели. В более сложных приложениях обильное число `@Published` полей приводит к фризам и лагам

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+    @Published var departureTimeIntervals: [DepartureTimeInterval] = []
+    @Published var hasTransfers: Bool = true
+    
+    private let navigationService = NavigationService.shared
```
Одна  `ObservableObject` `ViewModel` используется внутри другой `ObservableObject` `ViewModel`. Это не очень хороший паттерн. Вот [статья](https://rhonabwy.com/2021/02/13/nested-observable-objects-in-swiftui/) на этот счет

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+    }
+    
+    func setSearchButtonEnable() -> Bool {
+        return fromStation != nil && toStation != nil 
```
В однострочных выражениях `return` можно опускать

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+                    }
+                }
+            }
+        } catch ErrorsType.internetConnectError {
```
Можно оптимизировать и упростить код:
``` swift
var stationList: [Settlements] = []
let testSettlements = Set(["Москва", "Санкт-Петербург", "Сочи", "Горный Воздух", "Краснодар", "Казань", "Омск"])
do {
    let allStationsList = try await getStationsList()
    stationList = allStationsList.countries?
        .flatMap { $0.regions ?? [] }
        .flatMap { $0.settlements ?? [] }
        .filter { testSettlements.contains($0.title ?? "") } ?? []
}
```

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+        }
+        
+        allSettlements = stationList.filter { $0.title != ""}
+        isLoading = allSettlements.isEmpty ? true : false
```
`Bool` можно опустить:
`isLoading = allSettlements.isEmpty`

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+        
+        allSettlements = stationList.filter { $0.title != ""}
+        isLoading = allSettlements.isEmpty ? true : false
+        allSettlements.sort {$0.title! < $1.title! }
```
Давай нигде не будем использовать `force` операции

NaR0RoW commented on TravelSchedule/ViewModel/ScheduleViewModel.swift:
```diff
@@ -0,0 +1,212 @@
+        )
+        
+        return client
+    }
```
Эти методы стоит вынести в отдельный сервис. ViewModel в SwiftUI должна отвечать за отображение данных. Лучше сделать другой сервис, который будет отвечать за подготовку данных для ViewModel

NaR0RoW commented on TravelSchedule/Views/Schedule/CarrierCardView.swift:
```diff
@@ -0,0 +1,102 @@
+        startDateString = startDateFormatter.string(from: generalFormatter.date(from: segmentInfo.thread?.start_date ?? "") ?? currentDate)
+        beginTime = timeFormatter.string(from: generalFormatter.date(from: segmentInfo.departure ?? "") ?? currentDate)
+        endTime = timeFormatter.string(from: generalFormatter.date(from: segmentInfo.arrival ?? "") ?? currentDate)
+    }
```
View не должна заниматься настройкой данных для отображения. View должна получать уже готовые данные от сервиса / view model

NaR0RoW commented on TravelSchedule/Views/Schedule/CarrierCardView.swift:
```diff
@@ -0,0 +1,102 @@
+                    VStack(alignment: .leading) {
+                        Text(segment.thread?.carrier?.title ?? "")
+                            .font(.system(size: 17, weight: .regular))
+                        if segment.has_transfers ?? false {
```
Везде, где это возможно (если layout не изменяется), лучше использовать `opacity` вместо `if` для сокрытия `View`. [Вот хороший пример с ссылками на `WWDC`](https://stackoverflow.com/a/57420479)

NaR0RoW commented on TravelSchedule/Views/Schedule/CarrierCardView.swift:
```diff
@@ -0,0 +1,102 @@
+                .foregroundStyle(.ypBlackUniversal)
+            }
+            .padding([.leading, .trailing], 16)
+            .padding([.top, .bottom], 14)
```
`.leading + .trailing` == `.horizontal`
`.top +.bottom == .vertical`

Можно заменить на:
``` swift
.padding(.horizontal, 16)
.padding(.vertical, 14)
```

NaR0RoW commented on TravelSchedule/Views/Schedule/CarrierInfoView.swift:
```diff
@@ -0,0 +1,70 @@
+    
+    var body: some View {
+        ZStack {
+            Color.ypWhite.ignoresSafeArea(.all)
```
`.all` можно опустить, т.к. это значение по умолчанию

NaR0RoW commented on TravelSchedule/Views/Schedule/CarrierInfoView.swift:
```diff
@@ -0,0 +1,70 @@
+        ZStack {
+            Color.ypWhite.ignoresSafeArea(.all)
+            VStack(alignment: .leading) {
+                VStack(alignment: .center) {
```
`(alignment: .center)` можно опустить, т.к. это значение по умолчанию

NaR0RoW commented on TravelSchedule/Views/Schedule/CarriersView.swift:
```diff
@@ -0,0 +1,87 @@
+                                            .foregroundStyle(viewModel.isFilter ? .ypRedUniversal : .ypBlue)
+                                    }
+                                } icon: {}
+//                                    .padding()
```
Закомментированную строчку можно удалить

NaR0RoW commented on TravelSchedule/Views/Schedule/CarriersView.swift:
```diff
@@ -0,0 +1,87 @@
+            .foregroundStyle(.ypBlack)
+            .padding(16)
+            
+        } //ZStack
```
Стоит удалять закомментированный код

NaR0RoW commented on TravelSchedule/Views/Schedule/SearchBar.swift:
```diff
@@ -0,0 +1,76 @@
+struct SearchBar: View {
+    
+    @Binding var searchText: String
+    @State var isSearching: Bool = false
```
Все `@State` свойства рекомендую указывать как `private`, потому что они не должны быть видны за пределами `View`. Если нужно изменить значение внутри `View`, то тогда следует использовать `@Binding`

NaR0RoW commented:
Если станция не найдена, то отображается надпись "Город не найден":

https://github.com/user-attachments/assets/8256df0e-895c-47b0-8254-a2d151e9775b

NaR0RoW commented:
Почему-то случайным образом показываются станции в городе. На видео я всегда выбираю Москву, но видно, что отображаются станции из других городов:

https://github.com/user-attachments/assets/3adff558-fa42-4a56-9d56-d026bc3ee8a3

NaR0RoW commented:
Не получилось составить ни один маршрут. Либо маршрут просто не находится, либо возникает следующая ошибка:

`Unexpected response, expected status code: ok, response: undocumented(statusCode: 404, OpenAPIRuntime.UndocumentedPayload(headerFields: [(field: X-Content-Type-Options: nosniff, next: 65535), (field: Content-Type: application/json; charset=utf-8, next: 65535), (field: Content-Length: 305, next: 65535), (field: Date: Mon, 03 Mar 2025 18:12:45 GMT, next: 65535), (field: Allow: GET, next: 65535), (field: Vary: Accept, Accept-Encoding, next: 65535), (field: Content-Encoding: gzip, next: 65535)], body: Optional(OpenAPIRuntime.HTTPBody)))`

При этой ошибке возникает следующей баг с навигацией:

https://github.com/user-attachments/assets/fd73c6ea-9d7e-4f4b-995d-8c28de2bd6de

Lisiykod commented:
> Почему-то случайным образом показываются станции в городе. На видео я всегда выбираю Москву, но видно, что отображаются станции из других городов:
> Simulator.Screen.Recording.-.iPhone.16.Pro.-.2025-03-03.at.21.13.44.mp4

Привет!
только добралась до ноутбука.
спасибо большое за видео, проверю в чем дело.
видимо после финальных правок все сломалось, и при проверке "повезло" не отловить такое поведение(

NaR0RoW commented:
> > Почему-то случайным образом показываются станции в городе. На видео я всегда выбираю Москву, но видно, что отображаются станции из других городов:
> > Simulator.Screen.Recording.-.iPhone.16.Pro.-.2025-03-03.at.21.13.44.mp4
> 
> Привет! только добралась до ноутбука. спасибо большое за видео, проверю в чем дело. видимо после финальных правок все сломалось, и при проверке "повезло" не отловить такое поведение(

Привет!
Спасибо, что написала. Отправлю тогда свои комментарии к твоей работе, может сразу поправишь их вместе с фиксом основной проблемы

Lisiykod committed:
устранен баг с неправильным отображением станций,
скорректирован текст для отображения результата пустого поиска по станциям,
скорректирована верстка экрана с фильтрами,
добавлен фон для логотипов перевозчиков в экран с информацией о перевозчике.

Lisiykod commented:
> > > Почему-то случайным образом показываются станции в городе. На видео я всегда выбираю Москву, но видно, что отображаются станции из других городов:
> > > Simulator.Screen.Recording.-.iPhone.16.Pro.-.2025-03-03.at.21.13.44.mp4
> > 
> > 
> > Привет! только добралась до ноутбука. спасибо большое за видео, проверю в чем дело. видимо после финальных правок все сломалось, и при проверке "повезло" не отловить такое поведение(
> 
> Привет! Спасибо, что написала. Отправлю тогда свои комментарии к твоей работе, может сразу поправишь их вместе с фиксом основной проблемы

Супер, спасибо большое - столько всего полезного! постараюсь поправить по максимуму.

Lisiykod committed:
Устранен баг с выбрасыванием ошибки при отсутствии вариантов перевозок,
сетевой слой вынесен в провайдер,
форматирование даты вынесено во вью модель +
небольшие улучшения кода

Lisiykod commented:
Привет!
Спасибо за комментарии, все поправила. 
Только от синглтона в навигации еще не успела избавиться, надеюсь поправлю это в ближайшее время.

Lisiykod committed:
скорректировано отображение таб бара

NaR0RoW reviewed (approved):
Отличная работа 🔥

Lisiykod commented:
> Отличная работа 🔥

Спасибо! 💜
